### PR TITLE
feat: Add automated component/pipeline skeleton generator with comprehensive tooling

### PR DESCRIPTION
**Pull request by nsingla**
## Summary

  This PR introduces a complete skeleton generation system that dramatically reduces the friction for contributing new components and pipelines to the repository. Contributors can now create fully-structured, production-ready component/pipeline skeletons in seconds using simple make commands.

  ## 🎯 Problem Solved

  Previously, creating new components or pipelines required:
  - Manual creation of 8+ required files per component
  - Understanding complex directory structures and naming conventions
  - Writing boilerplate code, metadata, and documentation from scratch
  - Setting up proper test structures
  - Potential for inconsistencies and missing required files

  This created a high barrier to entry for new contributors and slowed down development.

  ## ✅ Solution Provided

  **Automated Skeleton Generation:**
  - `make component CATEGORY=data_processing NAME=my_component` - Creates complete component structure
  - `make pipeline CATEGORY=training NAME=my_pipeline` - Creates complete pipeline structure
  - `make tests TYPE=component CATEGORY=data_processing NAME=my_component` - Adds tests to existing skeletons
  - `make readme TYPE=component CATEGORY=data_processing NAME=my_component` - Auto-generates documentation

  **Key Features:**
  - 🏗️ **Complete file generation** - All 8+ required files created automatically with proper templates
  - 🧪 **Comprehensive testing** - Unit and integration test templates with proper fixtures
  - 📚 **Smart documentation** - Integration with existing README generator for auto-docs
  - ⚡ **Flexible workflow** - Option to skip tests initially (`NO_TESTS` flag) for rapid prototyping
  - 🎛️ **Make integration** - Consistent, easy-to-remember commands
  - ✅ **Quality assurance** - All generated code follows project standards and passes linting

  ## 📁 Files Created by Skeleton Generator

  ```text
  components/category/my_component/
  ├── __init__.py            # Import configuration
  ├── component.py           # Implementation template with TODOs
  ├── metadata.yaml          # Pre-configured metadata
  ├── README.md              # Documentation template
  ├── OWNERS                 # Maintainer template
  └── tests/                 # Test directory
      ├── __init__.py
      ├── test_component_unit.py     # Unit test template
      └── test_component_local.py    # Integration test template
```

 ## 🔧 Implementation Details

 ### Core Components:
  - scripts/generate_skeleton/generate_skeleton.py - Main skeleton generator (643 lines)
  - Helper functions for DRY code generation (_generate_core_files, _generate_test_files)
  - Comprehensive error handling with actionable user guidance
  - Support for both components and pipelines

 ### Makefile Integration:
  - component and pipeline targets with validation and NO_TESTS support
  - tests target for adding tests to existing skeletons
  - readme target integrated with existing README generator
  - Consistent argument patterns across all targets

 ### Quality Assurance:
  - 18 comprehensive unit tests with 100% functionality coverage
  - Tests for CLI interface, core functions, error handling, and edge cases
  - All code passes ruff linting and formatting standards
  - Integration with existing pre-commit hooks

 ## 📖 Documentation Improvements

  Added new "Quick Development Workflow" section to CONTRIBUTING.md featuring:
  - Overview table of all make commands
  - Step-by-step development process (6 clear steps)
  - Complete example workflow showing component creation to PR submission
  - Integration with existing quality tools (make format, make lint, pre-commit run)

 ## 🎛️ Usage Examples

  Create component with tests:
  `make component CATEGORY=data_processing NAME=csv_processor`

  Create pipeline without tests (rapid prototyping):
  `make pipeline CATEGORY=training NAME=model_trainer NO_TESTS`

  Add tests to existing component:
  `make tests TYPE=component CATEGORY=data_processing NAME=csv_processor`

  Generate documentation from code:
  `make readme TYPE=component CATEGORY=data_processing NAME=csv_processor`

  Complete workflow (skeleton → implementation → PR):
```text
  # 1. Create skeleton
  make component CATEGORY=data_processing NAME=csv_cleaner

  # 2. Implement logic in generated component.py file

  # 3. Generate docs  
  make readme TYPE=component CATEGORY=data_processing NAME=csv_cleaner

  # 4. Quality checks
  make format && make lint && pre-commit run

  # 5. Submit PR
  git add . && git commit -m "feat(data_processing): add csv_cleaner component"
```

 ## 🧪 Testing Coverage

  - 18 unit tests covering all functionality
  - CLI interface tests - argument validation, error handling, successful execution
  - Core function tests - file generation, skeleton creation, test addition
  - Error handling tests - missing directories, invalid names, conflicting options
  - Integration tests - end-to-end workflows with temporary directories

  All tests pass and provide comprehensive coverage of the skeleton generation system.

 ## 🎉 Benefits

  - ⚡ 10x faster component creation - Complete setup in seconds vs. 20+ minutes manually
  - 📏 Perfect consistency - All components follow identical structure and standards
  - 🎯 Lower barrier to entry - New contributors can start coding immediately
  - 🧪 Built-in best practices - Proper testing structure and documentation from day one
  - 🔧 Seamless integration - Works with all existing repository tooling and workflows
  - 📚 Self-documenting - Generated code includes comprehensive TODO guidance

 ## 🔍 Backwards Compatibility

  - ✅ No breaking changes to existing components or workflows
  - ✅ All existing make targets continue to work unchanged
  - ✅ New functionality is purely additive
  - ✅ Existing documentation structure preserved

  This PR transforms the contributor experience from manual, error-prone setup to automated, consistent, and fast skeleton generation, dramatically improving development velocity while maintaining high quality standards.

### Pre-Submission Checklist

- [x] All tests and CI checks pass
- [x] Pre-commit hooks pass without errors
- [x] You have [signed off your commits](https://www.kubeflow.org/docs/about/contributing/#sign-off-your-commits)
- [x] The title for your pull request (PR) should follow our title convention.
  [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### Additional Checklist Items for New or Updated Components/Pipelines

- [ ] `metadata.yaml` includes fresh `lastVerified` timestamp
- [ ] All [required files](https://github.com/kubeflow/pipelines-components/blob/main/docs/CONTRIBUTING.md#required-files)
  are present and complete
- [ ] OWNERS file lists appropriate maintainers
- [ ] README provides clear documentation with usage examples
- [ ] Component follows `snake_case` naming convention
- [ ] No security vulnerabilities in dependencies
- [ ] Containerfile included if using a custom base image

### CODE REVIEW

File: docs/CONTRIBUTING.md [hunk 1]
```diff
@@ -188,6 +188,160 @@ The `OWNERS` file enables code review automation by leveraging PROW commands:
...
+The following make targets simplify the development workflow:
+
+| Command | Description |
+|---------|-------------|
+| `make component CATEGORY=<cat> NAME=<name> [NO_TESTS]` | Create a new component skeleton |
+| `make pipeline CATEGORY=<cat> NAME=<name> [NO_TESTS]` | Create a new pipeline skeleton |
+| `make tests TYPE=<type> CATEGORY=<cat> NAME=<name>` | Add tests to existing component/pipeline |
+| `make readme TYPE=<type> CATEGORY=<cat> NAME=<name>` | Generate/update README from code |
+| `make format` | Auto-fix code formatting and linting issues |
+| `make lint` | Check code quality (formatting, linting, imports) |
```
Comment (hbelmiro) [inline docs/CONTRIBUTING.md]: nit: can we get this well formatted?

```suggestion
| Command                                                | Description                                       |
|--------------------------------------------------------|---------------------------------------------------|
| `make component CATEGORY=<cat> NAME=<name> [NO_TESTS]` | Create a new component skeleton                   |
| `make pipeline CATEGORY=<cat> NAME=<name> [NO_TESTS]`  | Create a new pipeline skeleton                    |
| `make tests TYPE=<type> CATEGORY=<cat> NAME=<name>`    | Add tests to existing component/pipeline          |
| `make readme TYPE=<type> CATEGORY=<cat> NAME=<name>`   | Generate/update README from code                  |
| `make format`                                          | Auto-fix code formatting and linting issues       |
| `make lint`                                            | Check code quality (formatting, linting, imports) |
```

File: scripts/generate_skeleton/generate_skeleton.py [hunk 2]
```diff
@@ -0,0 +1,638 @@
...
+- [ ] Add integration tests
+- [ ] Update documentation with actual parameters and usage
+- [ ] Update OWNERS file with actual maintainers
+- [ ] Add example usage notebooks
+"""
+
+    return files
+
+
+def _generate_test_files(skeleton_type: str, category: str, name: str):
```
Comment (hbelmiro) [inline scripts/generate_skeleton/generate_skeleton.py]: `category` is unused.

File: scripts/generate_skeleton/generate_skeleton.py [hunk 3]
```diff
@@ -0,0 +1,638 @@
...
+
+            next_steps = f"""
+Next steps:
+1. Update {created_dir}/OWNERS with your GitHub username
+2. Implement the logic in {created_dir}/{args.type}.py
+3. Update {created_dir}/metadata.yaml with correct dependencies and tags"""
+
+            if create_tests:
+                readme_cmd = f"make readme TYPE={args.type} CATEGORY={args.category} NAME={args.name}"
+                tests_cmd = f"make tests TYPE={args.type} CATEGORY={args.category} NAME={args.name}"
```
Comment (hbelmiro) [inline scripts/generate_skeleton/generate_skeleton.py]: `tests_cmd` is not used.

File: scripts/generate_skeleton/tests/test_cli.py [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+            original_cwd = Path.cwd()
+            import os
+
+            os.chdir(temp_dir)
+
+            try:
+                args = ["generate_skeleton.py", "--type=component", "--category=data_processing", "--name=my_processor"]
+
+                # Mock the working directory change to stay in temp_dir
+                def mock_chdir(path):
```
Comment (hbelmiro) [inline scripts/generate_skeleton/tests/test_cli.py]: `path` is unused.

```suggestion
                def mock_chdir(_path):
```

File: scripts/generate_skeleton/tests/test_cli.py [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+                args = [
+                    "generate_skeleton.py",
+                    "--type=pipeline",
+                    "--category=training",
+                    "--name=my_pipeline",
+                    "--no-tests",
+                ]
+
+                # Mock the working directory change to stay in temp_dir
+                def mock_chdir(path):
```
Comment (hbelmiro) [inline scripts/generate_skeleton/tests/test_cli.py]: `path` is unused.

```suggestion
                def mock_chdir(_path):
```

File: scripts/generate_skeleton/tests/test_cli.py [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+                args = [
+                    "generate_skeleton.py",
+                    "--type=component",
+                    "--category=data_processing",
+                    "--name=my_processor",
+                    "--tests-only",
+                ]
+
+                # Mock the working directory change to stay in temp_dir
+                def mock_chdir(path):
```
Comment (hbelmiro) [inline scripts/generate_skeleton/tests/test_cli.py]: `path` is unused.

```suggestion
                def mock_chdir(_path):
```

File: scripts/generate_skeleton/generate_skeleton.py [hunk 7]
```diff
@@ -0,0 +1,638 @@
...
+        """,
+    )
+
+    parser.add_argument("--type", choices=["component", "pipeline"], required=True, help="Type of skeleton to generate")
+
+    parser.add_argument(
+        "--category",
+        required=True,
+        help="Category for the component/pipeline (e.g., 'data_processing', 'training', 'ml_workflows')",
+    )
```
Comment (hbelmiro) [inline scripts/generate_skeleton/generate_skeleton.py]: The generator accepts arbitrary `--category`, but `pyproject.toml` packages only a fixed set of category modules. New categories may not be importable when installed. Suggest restricting `--category` to supported categories (existing dirs).

File: scripts/generate_skeleton/generate_skeleton.py [hunk 8]
```diff
@@ -0,0 +1,638 @@
+#!/usr/bin/env python3
+"""Kubeflow Pipelines Component/Pipeline Skeleton Generator
+
+This script creates skeleton component or pipeline structures based on the
+CONTRIBUTING.md guide. It generates all required files with proper templates.
+
+Usage:
+    python scripts/generate_skeleton.py --type=component --category=data_processing --name=my_processor
+    python scripts/generate_skeleton.py --type=pipeline --category=ml_workflows --name=my_training_pipeline
```
Comment (hbelmiro) [inline scripts/generate_skeleton/generate_skeleton.py]:
```suggestion
    python scripts/generate_skeleton/generate_skeleton.py --type=component --category=data_processing --name=my_processor
    python scripts/generate_skeleton/generate_skeleton.py --type=pipeline --category=ml_workflows --name=my_training_pipeline
```

**Review by hbelmiro — COMMENTED** [review]
@nsingla here's some initial thoughts so you can start working on.
I haven't done a complete review yet.

File: scripts/generate_skeleton/generate_skeleton.py [hunk 9]
```diff
@@ -0,0 +1,638 @@
...
+"""
+    else:  # pipeline
+        files["test_pipeline_unit.py"] = f"""\"\"\"Tests for the {name} pipeline.\"\"\"
+
+from unittest import mock
+
+from ..pipeline import {name}
+
+
+class Test{name.replace('_', '').title()}UnitTests:
```
Comment (hbelmiro) [inline scripts/generate_skeleton/generate_skeleton.py]: The class name transformation `name.replace('_', '').title()` produces incorrect casing.
For example, with `name = "my_processor"`:
Current result: `TestMyprocessorUnitTests`
Expected result: `TestMyProcessorUnitTests`
The issue is that `replace('_', '')` removes the underscore word boundaries before `.title()` is called, so `.title()` only sees `myprocessor` and capitalizes just the first letter.

Suggestion is change from:

```python
class Test{name.replace('_', '').title()}UnitTests:
```

To:

```python
class Test{name.title().replace('_', '')}UnitTests:
```

File: docs/CONTRIBUTING.md [hunk 10]
```diff
@@ -188,6 +188,160 @@ The `OWNERS` file enables code review automation by leveraging PROW commands:
 See [full Prow documentation](https://docs.prow.k8s.io/docs/components/plugins/approve/approvers/#lgtm-label)
 for usage details.
 
+## Quick Development Workflow
```
Comment (hbelmiro) [inline docs/CONTRIBUTING.md]: Having both "Quick Development Workflow" and "Development Workflow" as separate sections is confusing for contributors. They may not know which path to follow, or miss one section entirely.

Suggestion: Merge into a single "Development Workflow" section that presents both approaches (make commands vs. manual) as options within a unified linear flow. Use collapsible `<details>` blocks for the alternative approaches, so readers follow one path with choice points rather than choosing between two parallel workflows.
This also reduces duplication (e.g., both sections cover formatting, linting, and testing with similar content).

File: scripts/generate_skeleton/tests/test_generate_skeleton.py [hunk 11]
```diff
@@ -0,0 +1,240 @@
+"""Unit tests for generate_skeleton.py"""
+
+import tempfile
+from pathlib import Path
+
+import pytest
+
+from ..generate_skeleton import (
+    _generate_core_files,
+    _generate_test_files,
```
[the remark this answers is not in the document]
Comment (nsingla) [reply scripts/generate_skeleton/tests/test_generate_skeleton.py]: yeah, but its just for unit testing a script, so I think its fine, but if you strongly feel about it, I can switch the functions to pubic functions

[the remark this answers is not in the document]
Comment (hbelmiro) [reply scripts/generate_skeleton/tests/test_generate_skeleton.py]: I don't know. It doesn't feel right to test private functions. Switching to public functions would just be a hack.
If it's too much work, you can leave as it is and address it in a follow-up PR.

File: scripts/generate_skeleton/generate_skeleton.py [hunk 12]
```diff
@@ -0,0 +1,638 @@
...
+"""
+
+import argparse
+import os
+import sys
+from datetime import datetime, timezone
+from pathlib import Path
+
+
+def _generate_core_files(skeleton_type: str, category: str, name: str):
```
[the remark this answers is not in the document]
Comment (hbelmiro) [reply scripts/generate_skeleton/generate_skeleton.py]: Can we just remove it then? I don't see a point in having a lot of useless code. If you really need a placeholder, you can just drop a `#placeholder` comment.

**Review by hbelmiro — COMMENTED** [review]
Can you please remove the 🚀 emoji from the title so it follows the title convention?

File: scripts/generate_skeleton/generate_skeleton.py (hunk 9 above)
Comment (nsingla) [reply scripts/generate_skeleton/generate_skeleton.py]: Done

File: docs/CONTRIBUTING.md [hunk 13]
```diff
@@ -200,9 +200,68 @@ git fetch upstream
...
+The following make targets simplify the development workflow:
+
+| Command                                                | Description                                       |
+|--------------------------------------------------------|---------------------------------------------------|
+| `make component CATEGORY=<cat> NAME=<name> [NO_TESTS]` | Create a new component skeleton                   |
+| `make pipeline CATEGORY=<cat> NAME=<name> [NO_TESTS]`  | Create a new pipeline skeleton                    |
+| `make tests TYPE=<type> CATEGORY=<cat> NAME=<name>`    | Add tests to existing component/pipeline          |
+| `make readme TYPE=<type> CATEGORY=<cat> NAME=<name>`   | Generate/update README from code                  |
+| `make format`                                          | Auto-fix code formatting and linting issues       |
+| `make lint`                                            | Check code quality (formatting, linting, imports) |
```
Comment (hbelmiro) [inline docs/CONTRIBUTING.md]:
```suggestion
| Command                                                | Description                                       |
|--------------------------------------------------------|---------------------------------------------------|
| `make component CATEGORY=<cat> NAME=<name> [NO_TESTS]` | Create a new component skeleton                   |
| `make pipeline CATEGORY=<cat> NAME=<name> [NO_TESTS]`  | Create a new pipeline skeleton                    |
| `make tests TYPE=<type> CATEGORY=<cat> NAME=<name>`    | Add tests to existing component/pipeline          |
| `make readme TYPE=<type> CATEGORY=<cat> NAME=<name>`   | Generate/update README from code                  |
| `make format`                                          | Auto-fix code formatting and linting issues       |
| `make lint`                                            | Check code quality (formatting, linting, imports) |
```

File: scripts/generate_skeleton/tests/test_generate_skeleton.py [hunk 14]
```diff
@@ -0,0 +1,316 @@
...
+            "category with spaces",  # Spaces
+        ],
+    )
+    def testvalidate_category_invalid_cases(self, invalid_category):
+        """Test validate_category raises ValueError for invalid categories."""
+        with pytest.raises(ValueError):
+            validate_category(invalid_category)
+
+    def test_get_existing_categories_empty(self):
+        """Test _get_existing_categories with no existing directories."""
```
Comment (hbelmiro) [inline scripts/generate_skeleton/tests/test_generate_skeleton.py]:
```suggestion
        """Test get_existing_categories with no existing directories."""
```

File: scripts/generate_skeleton/generate_skeleton.py [hunk 15]
```diff
@@ -0,0 +1,440 @@
...
+    if "." in name:
+        raise ValueError("Name cannot contain dots (.)")
+
+    # Check for Python identifier validity
+    if not name.isidentifier():
+        raise ValueError(f"Name '{name}' is not a valid Python identifier")
+
+    # Check for Python keywords
+    if keyword.iskeyword(name):
+        raise ValueError(f"Name '{name}' is a Python keyword and cannot be used")
```
Comment (hbelmiro) [inline scripts/generate_skeleton/generate_skeleton.py]: `keyword.iskeyword()` only catches hard keywords. Consider also checking for soft keywords introduced in Python 3.10+ using `keyword.issoftkeyword()`:

```suggestion
    if keyword.iskeyword(name) or keyword.issoftkeyword(name):
        raise ValueError(f"Name '{name}' is a Python keyword and cannot be used")
```

Without this, users could create components named `match`, `case`, or `type` which are contextual keywords in modern Python.

File: scripts/generate_skeleton/generate_skeleton.py [hunk 16]
```diff
@@ -0,0 +1,440 @@
...
+
+    if "." in category:
+        raise ValueError("Category cannot contain dots (.)")
+
+    # Enforce snake_case (no uppercase letters, allow underscores)
+    if category != category.lower():
+        raise ValueError("Category must be lowercase (snake_case)")
+
+    # Allow letters, numbers, and underscores
+    if not category.replace("_", "").isalnum():
```
Comment (hbelmiro) [inline scripts/generate_skeleton/generate_skeleton.py]: `validate_category()` allows categories starting with numbers (e.g., `123abc`) because `isalnum()` doesn't enforce letter-first requirement. This could create invalid Python package names.

```suggestion
    if not category[0].isalpha():
```

File: scripts/generate_skeleton/generate_skeleton.py [hunk 17]
```diff
@@ -0,0 +1,440 @@
...
+def generate_core_files(skeleton_type: str, category: str, name: str):
+    """Generate core files for skeleton based on type.
+
+    Args:
+        skeleton_type: Type of skeleton ('component' or 'pipeline')
+        category: Category name
+        name: Skeleton name
+
+    Returns:
+        Dict of filename to content mappings
```
Comment (hbelmiro) [inline scripts/generate_skeleton/generate_skeleton.py]: This is not true.

File: scripts/generate_skeleton/tests/test_generate_skeleton.py [hunk 18]
```diff
@@ -0,0 +1,316 @@
...
+                with pytest.raises(ValueError) as exc_info:
+                    create_tests_only("component", "data_processing", "my_processor")
+
+                assert "missing main file" in str(exc_info.value)
+
+            finally:
+                os.chdir(original_cwd)
+
+
+class TestValidationFunctions:
```
[the remark this answers is not in the document]
Comment (nsingla) [reply scripts/generate_skeleton/tests/test_generate_skeleton.py]: it lost its underscore when I changed the private `_validname` function to a public one :doh

File: scripts/generate_skeleton/generate_skeleton.py [hunk 19]
```diff
@@ -0,0 +1,440 @@
...
+
+    # Prepare type-specific template variables
+    if skeleton_type == "component":
+        decorator_name = "component"
+        decorator_params = (
+            '    base_image="python:3.11",\n'
+            '    # packages_to_install=["numpy", "pandas"],  # Add your dependencies here'
+        )
+        param_default = ""
+        param_extras = ",\n    # Add your output artifacts here\n" "    # output_artifact: dsl.Output[dsl.Artifact]"
```
Comment (hbelmiro) [inline scripts/generate_skeleton/generate_skeleton.py]: This line is weird. Is the implicit concatenation intentional? How about using triple quotes to increase readability?

File: scripts/generate_skeleton/generate_skeleton.py [hunk 20]
```diff
@@ -0,0 +1,440 @@
...
+        param_default = ' = "default_value"'
+        param_extras = ""
+        return_annotation = ""
+        docstring_title = "TODO: Add pipeline description"
+        args_extra = ""
+        returns_documentation = "        Pipeline outputs or None if no outputs are needed."
+
+    # Consolidated function parameters and documentation
+    function_parameters = (
+        f"    # Add your {skeleton_type} parameters here\n" f"    input_param: str{param_default}{param_extras},"
```
Comment (hbelmiro) [inline scripts/generate_skeleton/generate_skeleton.py]: Same here about implicit concatenation.

**Review by hbelmiro — APPROVED** [review]
/lgtm
/approve
